### PR TITLE
Core: Rework multi date formatter merging

### PR DESCRIPTION
**Pull request by rjernst**
This commit moves the MergedDateFormatter to a package private class and
reworks joda DateFormatter instances to use that instead of a single
DateTimeFormatter with multiple parsers. This will allow the java and
joda multi formats to share the same format parsing method in a
followup.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 1]
```diff
@@ -1461,6 +1468,86 @@ private static DateFormatter forPattern(String input, Locale locale) {
...
     }
 
+    static class MergedDateFormatter implements DateFormatter {
+
+        private final String pattern;
+        private final List<DateFormatter> formatters;
+        private final List<DateMathParser> dateMathParsers;
+
+        MergedDateFormatter(String pattern, List<DateFormatter> formatters) {
+            assert formatters.size() > 0;
```
Comment (spinscale) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: the array was switched to a list while copying, I assume that was intentional?

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: Yes that was intentional

File: server/src/main/java/org/elasticsearch/common/time/DateFormatter.java [hunk 2]
```diff
@@ -86,7 +87,7 @@ default DateTime parseJoda(String input) {
      * Return the given millis-since-epoch formatted with this format.
      */
     default String formatMillis(long millis) {
-        return format(Instant.ofEpochMilli(millis));
+        return format(ZonedDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneOffset.UTC));
```
Comment (spinscale) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatter.java]: 👍

**Review by spinscale — APPROVED** [review]
LGTM - happy when CI is happy

**Comment by rjernst** [comment]
@elasticmachine test this please

**Pull request merged**